TypeScript-Definitionen aus Pydantic generieren

tumidi opened the pull request:
- Siehe Issue #191
- Hängt ab von questionpy-org/questionpy-server#155

### TypeScript-Definitionen generieren

Fügt ein `generate-ts-types.py`-Skript hinzu, welches automatisch TypeScript-Definitionen für das Vue-Frontend generiert. Dadurch haben wir eine Single-Source-of-Truth für das API-Schema und können auf duplizierte Schemata/Typdefinitionen im Frontend verzichten.

Das Skript generiert JSON Schemata aus ausgewählten Models, dataclasses, etc. und leitet diese an das `json-schema-to-types.ts`-Skript weiter, welches intern [json-schema-to-typescript](https://github.com/bcherny/json-schema-to-typescript) verwendet, um aus den Schemata TypeScript-Definitionen zu erstellen. json-schema-to-typescript ist in diesem Bereich das mit Abstand verbreitetste Tool, was ich finden konnte und liefert mit etwas Tweaking auch gute Ergebnisse.

Ansonsten gibt es hier Anpassungen an den in den Endpunkten exponierten Strukturen, bspw. habe ich `AttemptRenderData` zu einer `dataclass` refaktorisiert, für die sich ein Schema generieren lässt und exportiere `ClientQuestionDisplayOptions` direkt aus Python.

Größere Änderungen gab es hier nur für `questionpy_sdk.webserver.controllers.attempt.errors`, aber das habe ich schlussendlich in einen eigenen Commit ausgegliedert (siehe weiter unten).

#### Tweaks im Konvertierungsprozesses:

Damit die Ergebnisse auch brauchbar sind, habe ich ein paar Änderungen am Schema vorgenommen:

- Die Strukturierung der Model-Felder hat direkten Einfluss auf die Schema-Struktur und damit die Typen, d.h. es macht Sinn hier und da ein RootModel oder ein Typ-Alias einzuführen, damit separate Typ-Definitionen entstehen, anstatt dass diese inline ausgeschrieben werden. questionpy-org/questionpy-server#155 ist ein gutes Beispiel dafür.
- Felder mit default werden in TypeScript zu optionalen Feldern, was für die API-Typen aber keinen Sinn macht, da diese Felder ja immer ausgefüllt sind. (siehe `TypeScriptSchemaGenerator.model_schema`-Methode in `generate-ts-types.py`)
- json-schema-to-typescript erstellt für Properties, die einen Titel-Feld haben automatisch einen eigenen Typen bzw. Typ-Alias im Output, d.h. per default bekommt man eine große Zahl von sinnlosen Typ-Aliasen im Output, zB. `type Template1 = string`, `type Template2 = string`, usw. Ich habe eine gute Balance gefunden, indem ich das Titelfeld für alle Felder unterbinde, die nur aus primitiven Typen bestehen. (`TypeScriptSchemaGenerator.field_title_should_be_set`)

`json-schema-to-types.ts` jagt den generierten Output von json-schema-to-typescript schließlich noch einmal durch ESLint und Prettier. Der ESLint-Schritt führt alle Code-Fixes und -Suggestions aus, so dass der TS-Code letztendlich idiomatischer und einheitlicher ist. Das ist zwar aktuell nicht zwingend notwendig, aber ich habe im Verlauf generierte Typ-Variationen gefunden, die ESLint aufräumen und umformen möchte, daher macht das für etwaige zukünftige Schema-Updates schon Sinn, denke ich.

### Anpassung Frontend

Das Frontend macht keine Runtime-Validation der Daten mehr, d.h. Entfernung der Zod-Schemata und -Typen, Entfernung der Validierungslogik. Ein paar Convenience-Definitionen und Utilities, die noch übrig waren leben jetzt in `frontend/src/types/index.ts` bzw. `frontend/src/types/typeUtils.ts`.

### QuestionUI Render Errors

`questionpy_sdk/webserver/controllers/attempt/errors` war eh noch auf der TODO-Liste und musste für eine vollständige Typen-Generierung so verändert werden, dass Pydantic Schemata dafür ausspucken kann, was im Übrigen auch die JSON-Serialisierung vereinfacht, für die es vorher einen extra JSONEncoder brauchte.

- Benutzt ein Template-basiertes Mini-Format, welches bisher nicht strukturiert an das Frontend übergeben wurde. (Bisher wurden einfach String-Repräsentationen ausgegeben.) Habe die Logik hier nicht berührt, aber das Frontend so verändert, dass es diese Templates verarbeitet und entsprechend Werte in `code`-Tags ausgibt.
- `RenderErrorCollections` umbenannt nach `SectionErrorMap`, einfach weil ich nach dem zehnten Mal verwechseln mit `RenderErrorCollection` gedacht habe, das ist kein guter Name :smile: 

Außerdem `DetailedServerError` eingeführt. Das ist eine Struktur, um Backend-seitige Fehler an das Frontend zu kommunizieren. Unterstützt gerade nur string (Backend Stacktrace), oder Liste von Pydantic errors. Aber das ist als Vorbereitung zu sehen, so wie in https://github.com/questionpy-org/questionpy-server/issues/128#issuecomment-2587526618 beschrieben (Stacktraces serialisieren und schön anzeigen). Ggf. könnten wir das später noch aufbohren und eine Hierarchie `TracebackServerError`, `ValidationServerError`, `WorkerServerError`, usw. oder so ähnlich daraus machen.

tumidi committed:
feat(frontend): generate TypeScript definitions from Pydantic models

Add `generate-ts-types.py` to export JSON Schema from Pydantic models, and
`json-schema-to-types.ts` to convert the schema into TypeScript definitions.

Closes #191

tumidi committed:
feat(frontend): remove Zod and rely on generated TypeScript types

Switch to static type checking only. Runtime validation of API responses is no longer performed.

tumidi committed:
feat: improve structured error handling and display

Refactor render error handling into structured models and components.
Display detailed server-side errors in the frontend with improved UI and type support.

tumidi committed:
chore(frontend): avoid TSDoc-style headers in `.ts` files

Replaced `/**` with `/*` in file headers to prevent misinterpretation as TSDoc blocks when no
imports precede a declaration.

MHajoha commented on generate-ts-types.py:
```diff
@@ -0,0 +1,208 @@
+    | core_schema.UrlSchema
+    | core_schema.MultiHostUrlSchema
+    | core_schema.UuidSchema
+)
```
Ich bin überrascht, dass Mypy da nicht meckert, aber ich würde trotzdem <code><b><i>type</i></b> PrimitiveSchema = ...</code> schreiben

MHajoha commented on generate-ts-types.py:
```diff
@@ -0,0 +1,208 @@
+            return False
+
+        if is_nullable_field(schema):
+            return self.field_title_should_be_set(schema["schema"])
```
Macht `super().field_title_should_be_set(schema)` unten das nicht auch?

MHajoha commented on questionpy_sdk/webserver/controllers/attempt/question_ui.py:
```diff
@@ -187,12 +187,14 @@ class ClientQuestionDisplayOptions(BaseModel):
+            DisplayRole.SCORER,
+            DisplayRole.TEACHER,
+        }
+    )
```
Pydantic kopiert mutable defaults eigentlich

Edit: Beim rumprobieren ist mir aufgefallen, dass das `default` im Json-Schema den Generator dazu verleitet, identische Typen zu duplizieren. War das der Grund für die `default_factory`s? Dann wäre nur ein Kommentar irgendwo sweet.

tumidi replied:
> War das der Grund für die default_factorys?

Einige von denen waren zwischenzeitlich dataclasses, wo die `default_factory` ja notwendig ist. Das hat sich dann so eingemogelt, aber hast völlig recht.

MHajoha commented on questionpy_sdk/webserver/middlewares/error.py:
```diff
@@ -2,22 +2,41 @@
+        description=(
+            "Optional detailed error information, which may include a stack trace or Pydantic validation errors."
+        )
+    )
```
Angeblich kann Pydantic docstrings als description verwenden: https://docs.pydantic.dev/latest/api/config/#pydantic.config.ConfigDict.use_attribute_docstrings

Wenn das auch für dataclasses geht, wäre das natürlich cooler :)

MHajoha commented on questionpy_sdk/webserver/controllers/attempt/errors.py:
```diff
@@ -2,49 +2,64 @@
-    return opening + f"{closing}, {opening}".join(values) + f"{closing} and {last_value}"
+    def __getitem__(self, key: str) -> str | Sequence[str]:
+        return self.root[key]
 
+    def __setitem__(self, key: str, value: str | Sequence[str]) -> None:
+        self.root[key] = value
 
-def _element_representation(element: etree._Element) -> str:
-    # Return the whole element if it is a PI.
-    if isinstance(element, etree._ProcessingInstruction):
-        return str(element)
+    def __delitem__(self, key: str) -> None:
+        del self.root[key]
 
-    # Create the prefix of an element. We do not want to keep 'html' as a prefix.
-    prefix = f"{element.prefix}:" if element.prefix and element.prefix != "html" else ""
-    return prefix + etree.QName(element).localname
+    def __iter__(self) -> Iterator[str]:  # type: ignore[override]
+        return iter(self.root)
+
+    def __len__(self) -> int:
+        return len(self.root)
```
Hier tut es doch auch ein type alias, oder?

tumidi replied:
```py
type TemplateKwargs = Annotated[
    MutableMapping[str, str | Sequence[str]],
    Field(
        description="A mapping of placeholder keys to values for a template.",
        json_schema_extra={"default": {}},
    ),
]
```

Ich hatte dann letztendlich doch viel mit dem `RootModel` gemacht, da damit docstrings parsing funktioniert und man auch nicht das das schema tweaken muss. Aber ja, ist schon schöner und kürzer.

MHajoha commented on questionpy_sdk/webserver/controllers/attempt/errors.py:
```diff
@@ -169,145 +185,182 @@ def __init__(self, element: etree._Element, placeholder: str | None, available:
+    | DuplicateNameError
+    | XMLSyntaxError,
+    Field(discriminator="kind"),
+]
```
```suggestion
type Error = Annotated[
    InvalidAttributeValueError
    | ConversionError
    | PlaceholderReferenceError
    | InvalidCleanOptionError
    | InvalidContentError
    | ExpectedAncestorError
    | UnknownElementError
    | UnknownAttributeError
    | DuplicateNameError
    | XMLSyntaxError,
    Field(discriminator="kind"),
]
```

MHajoha commented on questionpy_sdk/webserver/controllers/attempt/errors.py:
```diff
@@ -169,145 +185,182 @@ def __init__(self, element: etree._Element, placeholder: str | None, available:
 
+ErrorSection = Annotated[
+    Literal["formulation", "general_feedback", "specific_feedback", "right_answer"], Field(title="ErrorSection")
+]
```
`type`

MHajoha commented on questionpy_sdk/webserver/controllers/attempt/errors.py:
```diff
@@ -169,145 +185,182 @@ def __init__(self, element: etree._Element, placeholder: str | None, available:
+        return iter(self.root)
+
+    def __len__(self) -> int:
+        return len(self.root)
```
Auch hier scheint ein type alias zu reichen

MHajoha commented on frontend/src/components/attempt/AttemptRenderErrors/ErrorTemplate.vue:
```diff
@@ -0,0 +1,84 @@
+    const parts: Part[] = []
+    let lastIndex = 0
+
+    template.replace(/{(\w+)}/g, (match, key, offset) => {
```
Ist das ein Pattern, `replace` dafür zu verwenden? Sonst lieber `matchAll`, das scheint mir weniger verwirrend.

Beide Funktionen könnten einige Kommentare vertragen, es hat mich eine Weile gebraucht, zu verstehen, was hier passiert.

Gibt es in Vue keine gute Möglichkeit, direkt programmatisch Text und `<code>`-Elemente aneinanderzureihen?

tumidi replied:
Ich habe noch mal nachgeschraubt, aber `String.split()`, statt `matchAll()` verwendet, was die manuelle Index-Verwaltung überflüssig macht, Kommentare geschrieben und `h()` verwendet: 0be3e7707f674c6ebf917c40f0dcfe31197c8a80

Sag gerne, ob das cooler ist so.

MHajoha commented on frontend/src/components/attempt/AttemptRenderErrors/ErrorTemplate.vue:
```diff
@@ -0,0 +1,84 @@
+        parts.push({ type: 'text', content: i < restValues.length - 1 ? ', ' : ' and ' })
+    }
+
+    parts.push({ type: 'value', content: lastValue })
```
Das ist ziemlich schwer verständlich, insb. aufgrund der doppelten `reverse`.

Beispielhafter Gegenvorschlag
```typescript
    for (let i = 0; i < arrValue.length - 1; i++) {
        // Add all but the last value,
        parts.push({ type: 'value', content: arrValue[i] })
        // each followed by a ', ', except for the second-to-last one, which is followed by ' and '.
        parts.push({ type: 'text', content: i < arrValue.length - 1 ? ', ' : ' and ' })
    }

    // Add the last (or only) value.
    const lastValue = arrValue[arrValue.length - 1];
    parts.push({ type: 'value', content: lastValue })
```

tumidi replied:
auch in 0be3e7707f674c6ebf917c40f0dcfe31197c8a80

tumidi committed:
refactor(frontend): switch to render function for clearer placeholder handling

Replaced template loop with Vue render function for better readability. Used `String.split()` with
regex to avoid manual index tracking. Improved intent clarity and added inline documentation.
